daft/sql/sql_connection.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+
+import logging
+import warnings
+from typing import TYPE_CHECKING, Callable
+from urllib.parse import urlparse
+
+import pyarrow as pa
+
+if TYPE_CHECKING:
+    from sqlalchemy.engine import Connection
+
+logger = logging.getLogger(__name__)
+
+
+class SQLConnection:
+    def __init__(self, conn: str | Callable[[], Connection], driver: str, dialect: str) -> None:
+        self.conn = conn
+        self.dialect = dialect
+        self.driver = driver
+
+    def __repr__(self) -> str:
+        return f"SQLConnection(conn={self.conn})"
+
+    @classmethod
+    def from_url(cls, url: str) -> SQLConnection:
+        scheme = urlparse(url).scheme.strip().lower()
+        if "+" in scheme:
+            dialect, driver = scheme.split("+")
+        else:
+            dialect, driver = scheme, ""
+        return SQLConnection(url, driver, dialect)
+
+    @classmethod
+    def from_connection_factory(cls, conn_factory: Callable[[], Connection]) -> SQLConnection:
+        try:
+            with conn_factory() as connection:
+                if not hasattr(connection, "engine"):
+                    raise ValueError("The connection factory must return a SQLAlchemy connection object.")
+                dialect = connection.engine.dialect.name
+                driver = connection.engine.driver
+            return SQLConnection(conn_factory, driver, dialect)
+        except Exception as e:
+            raise ValueError(f"Unexpected error while calling the connection factory: {e}") from e
+
+    def read(
+        self, sql: str, projection: list[str] | None = None, predicate: str | None = None, limit: int | None = None
+    ) -> pa.Table:
+        sql = self._construct_sql_query(sql, projection, predicate, limit)
+        try:
+            return self._execute_sql_query(sql)
+        except RuntimeError as e:
+            if limit is not None:
```

Comment:
Oh dang, what is the use-case for retrying without a `limit`? Sounds pretty expensive.

Reply:
some dbs don't support limit: https://stackoverflow.com/questions/2832013/can-you-name-a-single-popular-database-that-doesnt-support-limit-statement 

however this shouldn't be necessary once we use a sql generation library to help build our queries.